### PR TITLE
Refactoring of CAST FORMAT (#7881)

### PR DESCRIPTION
**Pull request by TreeHunter9**
I look back at my implementation of CAST FORMAT, and I think it's kinda mess, it's trying to do parsing, validation and execution at the same time. In addition, it bloats the cvt.cpp file.   
In this implementation, I moved CAST FORMAT logic to a separate file, and tried to separate parsing, validation and execution steps, so the code will look less messy (I hope so).
Also, I add more FORMAT validation based on SQL standard.

### CODE REVIEW

File: src/include/firebird/impl/msg/jrd.h [hunk 1]
```diff
@@ -984,3 +984,4 @@ FB_IMPL_MSG(JRD, 981, trailing_part_of_string, -901, "HY", "000", "There is a tr
 FB_IMPL_MSG(JRD, 982, pattern_cant_be_used_without_other_pattern, -901, "HY", "000", "@1 can't be used without @2")
 FB_IMPL_MSG(JRD, 983, pattern_cant_be_used_without_other_pattern_and_vice_versa, -901, "HY", "000", "@1 can't be used without @2 and vice versa")
 FB_IMPL_MSG(JRD, 984, incompatible_format_patterns, -901, "HY", "000", "@1 incompatible with @2")
+FB_IMPL_MSG(JRD, 985, only_one_pattern_can_be_used, -901, "HY", "000", "Can use only one of this patterns @1")
```
Comment (asfernandes) [inline src/include/firebird/impl/msg/jrd.h:987]: one of these?

File: src/common/CvtFormat.cpp [hunk 2]
```diff
@@ -0,0 +1,1678 @@
...
+
+	InitInstance<TimeZoneTrie> timeZoneTrie;
+
+
+	#define CVT_FORMAT(id, format) constexpr Patterns format = 1llu << id - 1;
+	#define CVT_FORMAT2(id, format1, format2) constexpr Patterns format2 = 1llu << id - 1;
+	#define CVT_FORMAT_FLAG(id, format) constexpr Patterns format = 1llu << id - 1;
+	namespace Format
+	{
+		typedef long long unsigned Patterns;
```
Comment (asfernandes) [inline src/common/CvtFormat.cpp:149]: Use SINT64 or FB_UINT64.
Or the std variants like uint64_t.

File: src/common/CvtFormat.cpp [hunk 3]
```diff
@@ -0,0 +1,1678 @@
...
+	{
+		cb->err(Arg::Gds(isc_invalid_date_format) << string(pattern.data(), pattern.length()));
+	}
+
+	void incompatibleDateFormatException(std::string_view pattern, Callbacks* cb)
+	{
+		cb->err(Arg::Gds(isc_incompatible_date_format_with_current_date_type) << string(pattern.data(), pattern.length()));
+	}
+
+//---------------------------------------------------------------------------------------------------------------------
```
Comment (asfernandes) [inline src/common/CvtFormat.cpp:494]: Please indent comments.

File: src/common/CvtFormat.cpp [hunk 4]
```diff
@@ -0,0 +1,1678 @@
...
+		for (FB_SIZE_T i = 0; i < formatUpper.length(); i++)
+		{
+			const char symbol = formatUpper[i];
+			if (symbol != '\"')
+			{
+				formatUpper[i] = toupper(symbol);
+				continue;
+			}
+
+			while(true)
```
Comment (asfernandes) [inline src/common/CvtFormat.cpp:546]:
```suggestion
			while (true)
```

File: src/common/CvtFormat.cpp [hunk 5]
```diff
@@ -0,0 +1,1678 @@
...
+			{
+				const FB_SIZE_T pos = formatUpper.find('\"', i + 1);
+				if (pos == string::npos)
+					cb->err(Arg::Gds(isc_invalid_raw_string_in_date_format));
+				FB_SIZE_T tempPos = pos;
+				if (formatUpper[--tempPos] == '\\')
+				{
+					int backslashCount = 1;
+					while(formatUpper[--tempPos] == '\\')
```
Comment (asfernandes) [inline src/common/CvtFormat.cpp:555]:
```suggestion
					while (formatUpper[--tempPos] == '\\')
```

File: src/common/CvtFormat.cpp [hunk 6]
```diff
@@ -0,0 +1,1678 @@
...
+		return tokens;
+	}
+
+	string intToRoman(int num)
+	{
+		constexpr const char* const symbols[] = {"M", "CM", "D", "CD", "C", "XC", "L", "XL", "X", "IX", "V", "IV", "I"};
+		constexpr int values[] = {1000, 900, 500, 400, 100, 90, 50, 40, 10, 9, 5, 4, 1};
+
+		string roman;
+		for (int i = 0; i < 13; i++)
```
Comment (asfernandes) [inline src/common/CvtFormat.cpp:628]: Please use `FB_NELEM`.

Comment (asfernandes) [reply src/common/CvtFormat.cpp:628]: Or even better, single array with `std::pair` and "for-each".

File: src/common/CvtFormat.cpp [hunk 7]
```diff
@@ -0,0 +1,1678 @@
...
+				hours -= 12;
+		}
+		else
+		{
+			period = FormatStr::AM;
+			if (hours == 0)
+				hours = 12;
+		}
+
+		return { hours, period };
```
Comment (asfernandes) [inline src/common/CvtFormat.cpp:655]:
```suggestion
		return {hours, period};
```

File: src/common/CvtFormat.cpp [hunk 8]
```diff
@@ -0,0 +1,1678 @@
...
+		}
+
+		return string(timezoneBuffer, length);
+	}
+
+	string processDateTimeToStringTokens(const dsc* desc, const std::vector<Token>& tokens, const struct tm& times, int fractions, Callbacks* cb)
+	{
+		string result;
+
+		auto begin = tokens.begin();
```
Comment (asfernandes) [inline src/common/CvtFormat.cpp:705]:
```suggestion
		const auto begin = tokens.begin();
```

File: src/common/CvtFormat.cpp [hunk 9]
```diff
@@ -0,0 +1,1678 @@
...
+					invalidPatternException(it->patternStr, cb);
+			}
+
+			result += patternResult;
+		}
+
+		return result;
+	}
+
+//---------------------------------------------------------------------------------------------------------------------
```
Comment (asfernandes) [inline src/common/CvtFormat.cpp:929]: Please indent comments.

File: src/common/CvtFormat.cpp [hunk 10]
```diff
@@ -0,0 +1,1678 @@
...
+		{
+			timestampTZ.time_zone = TimeZoneUtil::makeFromOffset(sign(cvtData.timezoneOffsetInMinutes),
+				abs(cvtData.timezoneOffsetInMinutes) / 60, abs(cvtData.timezoneOffsetInMinutes) % 60);
+		}
+
+		return timestampTZ;
+	}
+}
+
+string CVT_datetime_to_format_string(const dsc* desc, const string& format, Callbacks* cb)
```
Comment (asfernandes) [inline src/common/CvtFormat.cpp:1621]: It looks a bit weird that these public functions are declared in `cvt.h` and `CvtFormats.h` is not a header to be included by users of this new "module". I would instead rename current 'CvtFormat.h` to `CvtFormatImpl.h` and declare the public functions in `CvtFormat.h` (with `CVT_format_` prefix).

File: src/common/CvtFormat.cpp (hunk 3 above)
Comment (TreeHunter9) [reply src/common/CvtFormat.cpp:494]: I'm not quiet get it, you mean `// ----`?

**Comment by asfernandes** [comment]
New files should be added to the Windows build.

**Pull request merged**